File: oracle/full_tree_data_observations.py
"""Extract bounded DWARF globals and aggregate-layout observations."""

from __future__ import annotations

import hashlib
import json
from pathlib import Path
import threading
from typing import Any

from oracle.bounded_shards import ShardInput
from oracle.full_tree_function_observations import _declaration
from oracle.full_tree_scope import canonical_json_bytes
from oracle.function_recovery_oracle import _attribute_chain, _decode_dwarf_string


class FullTreeDataObservationError(ValueError):
    """Raised when global/type evidence cannot be represented exactly."""


POLICY = {
    "id": "full-tree-data-observations",
    "version": 1,
    "globals": "static-storage-or-linkage-bearing-dwarf-variables",
    "types": "class-struct-union-enum-definitions-and-declarations",
}

TYPE_TAGS = {
    "DW_TAG_class_type": "class",
    "DW_TAG_structure_type": "struct",
    "DW_TAG_union_type": "union",
    "DW_TAG_enumeration_type": "enum",
}


def _configuration_sha256() -> str:
    schema = Path(__file__).with_name("full-tree-data-observations.schema.json").read_bytes()
    return hashlib.sha256(canonical_json_bytes(POLICY) + schema).hexdigest()


def data_shard_inputs(inventory: dict[str, Any], *, scope_sha256: str, rich_sha256: str) -> tuple[list[ShardInput], dict[str, list[dict[str, Any]]]]:
    units = {unit["id"]: unit for unit in inventory["units"]}
    inputs = []
    grouped = {}
    for shard in inventory["shards"]:
        records = [units[item] for item in shard["unitIds"]]
        payload = canonical_json_bytes(
            {"configurationSha256": _configuration_sha256(), "inventoryIndexSha256": inventory["indexSha256"], "richArtifactSha256": rich_sha256, "scopeSha256": scope_sha256, "shardId": shard["id"], "units": records}
        )
        inputs.append(ShardInput(shard["id"], hashlib.sha256(payload).hexdigest()))
        grouped[shard["id"]] = records
    return inputs, grouped


def _integer_attribute(die: Any, name: str) -> int | None:
    for source in _attribute_chain(die):
        attribute = source.attributes.get(name)
        if attribute is not None and isinstance(attribute.value, int) and not isinstance(attribute.value, bool):
            return int(attribute.value)
    return None


def _name(die: Any) -> str | None:
    attribute = die.attributes.get("DW_AT_name")
    return _decode_dwarf_string(attribute.value, f"DIE {hex(int(die.offset))} name", 16_384) if attribute else None


def _names(die: Any) -> list[str]:
    result = set()
    for source in _attribute_chain(die):
        for key in ("DW_AT_linkage_name", "DW_AT_MIPS_linkage_name", "DW_AT_name"):
            attribute = source.attributes.get(key)
            if attribute is not None:
                result.add(_decode_dwarf_string(attribute.value, f"DIE {hex(int(die.offset))} {key}", 16_384))
    return sorted(result)


def _type_reference(die: Any) -> tuple[str | None, Any | None]:
    if "DW_AT_type" not in die.attributes:
        return None, None
    target = die.get_DIE_from_attribute("DW_AT_type")
    return hex(int(target.offset)), target


def _location(die: Any, dwarf: Any, image_base: int) -> tuple[int | None, bool]:
    attribute = die.attributes.get("DW_AT_location")
    if attribute is None or not isinstance(attribute.value, bytes):
        return None, False
    from elftools.dwarf.dwarf_expr import DWARFExprParser  # type: ignore[import-untyped]
    operations = DWARFExprParser(die.cu.structs).parse_expr(attribute.value)
    tls = any(operation.op_name in {"DW_OP_form_tls_address", "DW_OP_GNU_push_tls_address"} for operation in operations)
    if len(operations) == 1 and operations[0].op_name == "DW_OP_addr":
        address = int(operations[0].args[0])
        return address - image_base if address >= image_base else None, tls
    return None, tls


def _member(child: Any) -> dict[str, Any]:
    kind = {"DW_TAG_member": "field", "DW_TAG_inheritance": "base", "DW_TAG_enumerator": "enumerator"}[child.tag]
    reference, _ = _type_reference(child)
    byte_offset = _integer_attribute(child, "DW_AT_data_member_location")
    return {
        "bitOffset": _integer_attribute(child, "DW_AT_data_bit_offset") or _integer_attribute(child, "DW_AT_bit_offset"),
        "bitSize": _integer_attribute(child, "DW_AT_bit_size"),
        "byteOffset": byte_offset,
        "kind": kind,
        "name": _name(child),
        "typeDieOffset": reference,
        "value": _integer_attribute(child, "DW_AT_const_value") if kind == "enumerator" else None,
        "virtuality": _integer_attribute(child, "DW_AT_virtuality") if kind == "base" else None,
    }


def validate_data_observation_shard(document: dict[str, Any], *, scope: dict[str, Any], scope_sha256: str, inventory: dict[str, Any], shard: ShardInput, units: list[dict[str, Any]]) -> None:
    try:
        import fastjsonschema  # type: ignore[import-untyped]
        schema = json.loads(Path(__file__).with_name("full-tree-data-observations.schema.json").read_text(encoding="utf-8"))
        fastjsonschema.compile(schema)(document)
    except Exception as error:
        raise FullTreeDataObservationError(f"data observations fail validation: {error}") from error
    expected_oracle = {"configurationSha256": _configuration_sha256(), "inventoryIndexSha256": inventory["indexSha256"], "richArtifactSha256": scope["oracle"]["richArtifactSha256"], "scopeSha256": scope_sha256}
    if document["oracle"] != expected_oracle or document["shard"] != {"id": shard.identifier, "inputSha256": shard.input_sha256}:
        raise FullTreeDataObservationError("data observation bindings do not match")
    if document["globals"] != sorted(document["globals"], key=lambda item: item["id"]) or document["types"] != sorted(document["types"], key=lambda item: item["id"]):
        raise FullTreeDataObservationError("data observations are not ordered")
    unit_ids = {unit["id"] for unit in units}
    if any(item["unitId"] not in unit_ids for item in document["globals"] + document["types"]):
        raise FullTreeDataObservationError("data observation owner is outside its shard")
    expected = {
        "bases": sum(member["kind"] == "base" for item in document["types"] for member in item["members"]),
        "enumerators": sum(member["kind"] == "enumerator" for item in document["types"] for member in item["members"]),
        "fields": sum(member["kind"] == "field" for item in document["types"] for member in item["members"]),
        "globals": len(document["globals"]), "scannedDies": document["counts"]["scannedDies"],
        "types": len(document["types"]), "units": len(units),
    }
    if document["counts"] != expected:
        raise FullTreeDataObservationError("data observation counts do not reconcile")


def produce_data_observation_shard(rich_artifact: Path, *, scope: dict[str, Any], scope_sha256: str, inventory: dict[str, Any], shard: ShardInput, units: list[dict[str, Any]], output: Path, cancelled: threading.Event) -> int:
    from elftools.elf.elffile import ELFFile  # type: ignore[import-untyped]
    stream = rich_artifact.open("rb")
    try:
        elf = ELFFile(stream)
        dwarf = elf.get_dwarf_info(follow_links=False)
        loads = [segment for segment in elf.iter_segments() if segment["p_type"] == "PT_LOAD" and int(segment["p_memsz"]) > 0]
        image_base = min(int(segment["p_vaddr"]) for segment in loads)
        globals_ = []
        types = []
        scanned = 0
        for unit in units:
            cu = dwarf.get_CU_at(int(unit["dwarfOffset"], 16))
            for die in cu.iter_DIEs():
                if cancelled.is_set():
                    raise FullTreeDataObservationError(f"data shard {shard.identifier} was cancelled")
                scanned += 1
                if die.tag in TYPE_TAGS:
                    members = [_member(child) for child in die.iter_children() if child.tag in {"DW_TAG_member", "DW_TAG_inheritance", "DW_TAG_enumerator"}]
                    identity = hashlib.sha256(f"{unit['id']}:{hex(int(die.offset))}".encode()).hexdigest()[:32]
                    types.append({"alignment": _integer_attribute(die, "DW_AT_alignment"), "byteSize": _integer_attribute(die, "DW_AT_byte_size"), "declaration": _declaration(die, dwarf, scope, unit), "declarationOnly": bool(_integer_attribute(die, "DW_AT_declaration") or 0), "dieOffset": hex(int(die.offset)), "id": f"type-observation-{identity}", "members": members, "name": _name(die), "tag": TYPE_TAGS[die.tag], "unitId": unit["id"]})
                elif die.tag == "DW_TAG_variable":
                    names = _names(die)
                    parent = die.get_parent()
                    external = bool(_integer_attribute(die, "DW_AT_external") or 0)
                    address, tls = _location(die, dwarf, image_base)
                    if not names or not (
                        external
                        or parent is None
                        or parent.tag in {"DW_TAG_compile_unit", "DW_TAG_namespace"}
                        or any(name.startswith("_Z") for name in names)
                        or address is not None
                        or tls
                    ):
                        continue
                    type_offset, type_die = _type_reference(die)
                    identity = hashlib.sha256(f"{unit['id']}:{hex(int(die.offset))}".encode()).hexdigest()[:32]
                    visibility = {1: "local", 2: "exported", 3: "qualified"}.get(_integer_attribute(die, "DW_AT_visibility"), "default" if external else "unknown")
                    globals_.append({"addressRva": hex(address) if address is not None else None, "alignment": _integer_attribute(type_die, "DW_AT_alignment") if type_die else None, "declaration": _declaration(die, dwarf, scope, unit), "dieOffset": hex(int(die.offset)), "external": external, "id": f"global-observation-{identity}", "mutability": "constant" if "DW_AT_const_value" in die.attributes else "mutable" if address is not None or tls else "unknown", "names": names, "reasonCode": None if address is not None else "tls-no-image-rva" if tls else "optimized-out-or-nonaddress-location", "size": _integer_attribute(type_die, "DW_AT_byte_size") if type_die else None, "tls": tls, "typeDieOffset": type_offset, "unitId": unit["id"], "visibility": visibility})
            dwarf._cu_cache.clear(); dwarf._cu_offsets_map.clear(); dwarf._linetable_cache.clear()
        globals_.sort(key=lambda item: item["id"]); types.sort(key=lambda item: item["id"])
        counts = {"bases": sum(member["kind"] == "base" for item in types for member in item["members"]), "enumerators": sum(member["kind"] == "enumerator" for item in types for member in item["members"]), "fields": sum(member["kind"] == "field" for item in types for member in item["members"]), "globals": len(globals_), "scannedDies": scanned, "types": len(types), "units": len(units)}
        document = {"counts": counts, "globals": globals_, "oracle": {"configurationSha256": _configuration_sha256(), "inventoryIndexSha256": inventory["indexSha256"], "richArtifactSha256": scope["oracle"]["richArtifactSha256"], "scopeSha256": scope_sha256}, "schemaVersion": 1, "shard": {"id": shard.identifier, "inputSha256": shard.input_sha256}, "types": types}
        validate_data_observation_shard(document, scope=scope, scope_sha256=scope_sha256, inventory=inventory, shard=shard, units=units)
        output.write_bytes(canonical_json_bytes(document))
        return len(globals_) + len(types)
    finally:
        stream.close()

File: src/main/kotlin/decompengine/oracle/fulltree/FullTreeDataObservations.kt
package decompengine.oracle.fulltree

import decompengine.oracle.core.OracleArtifacts
import decompengine.oracle.core.OracleJson
import decompengine.oracle.core.OracleSchemas
import java.util.Collections
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.longOrNull

class FullTreeDataObservationShardInput internal constructor(
    val identifier: String,
    val inputSha256: String,
    units: List<JsonObject>,
) {
    val units: List<JsonObject> = Collections.unmodifiableList(ArrayList(units))

    init {
        require(identifier.matches(Regex("[a-z0-9]+(?:-[a-z0-9]+)*"))) {
            "invalid full-tree data observation shard identifier"
        }
        require(inputSha256.matches(Regex("[0-9a-f]{64}"))) {
            "invalid full-tree data observation input digest"
        }
        require(this.units.isNotEmpty()) { "full-tree data observation shard has no units" }
    }
}

/** Kotlin-owned bindings and semantic validation for data-observation shards. */
object FullTreeDataObservations {
    val configurationSha256: String by lazy {
        OracleSchemas.configurationSha256("full-tree-data-observations", POLICY)
    }

    fun shardInputs(
        inventory: JsonObject,
        scopeSha256: String,
        richArtifactSha256: String,
    ): List<FullTreeDataObservationShardInput> {
        requireSha256(scopeSha256, "scope")
        requireSha256(richArtifactSha256, "rich artifact")
        val inventoryIndexSha256 = inventory.requiredString("indexSha256")
        requireSha256(inventoryIndexSha256, "inventory index")
        val unitRecords = inventory.requiredArray("units")
            .mapIndexed { index, value -> value.requiredObject("inventory unit $index") }
        if (unitRecords.map { it.requiredString("id") }.distinct().size != unitRecords.size) {
            throw FullTreeDataTruthException("inventory data observation unit identifiers are not unique")
        }
        val units = unitRecords.associateBy { it.requiredString("id") }
        val inputs = inventory.requiredArray("shards").mapIndexed { index, value ->
            val shard = value.requiredObject("inventory shard $index")
            val identifier = shard.requiredString("id")
            val records = shard.requiredArray("unitIds").map { unitId ->
                val id = unitId.requiredString("inventory shard unit ID")
                units[id] ?: throw FullTreeDataTruthException(
                    "inventory shard $identifier references unknown unit $id",
                )
            }
            if (records.map { it.requiredString("id") }.distinct().size != records.size) {
                throw FullTreeDataTruthException("inventory shard $identifier repeats a unit")
            }
            val payload = JsonObject(
                mapOf(
                    "configurationSha256" to JsonPrimitive(configurationSha256),
                    "inventoryIndexSha256" to JsonPrimitive(inventoryIndexSha256),
                    "richArtifactSha256" to JsonPrimitive(richArtifactSha256),
                    "scopeSha256" to JsonPrimitive(scopeSha256),
                    "shardId" to JsonPrimitive(identifier),
                    "units" to JsonArray(records),
                ),
            )
            FullTreeDataObservationShardInput(
                identifier,
                OracleArtifacts.sha256(OracleJson.canonicalBytes(payload)),
                records,
            )
        }
        if (inputs.isEmpty()) throw FullTreeDataTruthException("inventory has no data observation shards")
        if (inputs.map { it.identifier }.distinct().size != inputs.size) {
            throw FullTreeDataTruthException("inventory data observation shard identifiers are not unique")
        }
        return inputs
    }

    fun validateShard(
        document: JsonObject,
        scope: JsonObject,
        scopeSha256: String,
        inventory: JsonObject,
        shard: FullTreeDataObservationShardInput,
    ) {
        try {
            OracleSchemas.validate("full-tree-data-observations", document)
        } catch (failure: Exception) {
            throw FullTreeDataTruthException("data observations fail schema validation: ${failure.message}", failure)
        }
        val richArtifactSha256 = scope.requiredObject("oracle").requiredString("richArtifactSha256")
        val authenticatedShard = shardInputs(inventory, scopeSha256, richArtifactSha256)
            .singleOrNull { it.identifier == shard.identifier }
            ?: throw FullTreeDataTruthException("data observation shard is outside the authenticated inventory")
        if (
            shard.inputSha256 != authenticatedShard.inputSha256 ||
            shard.units != authenticatedShard.units
        ) {
            throw FullTreeDataTruthException("data observation shard input is not authenticated")
        }
        val expectedOracle = JsonObject(
            mapOf(
                "configurationSha256" to JsonPrimitive(configurationSha256),
                "inventoryIndexSha256" to JsonPrimitive(inventory.requiredString("indexSha256")),
                "richArtifactSha256" to JsonPrimitive(richArtifactSha256),
                "scopeSha256" to JsonPrimitive(scopeSha256),
            ),
        )
        val expectedShard = JsonObject(
            mapOf(
                "id" to JsonPrimitive(shard.identifier),
                "inputSha256" to JsonPrimitive(shard.inputSha256),
            ),
        )
        if (document.requiredObject("oracle") != expectedOracle || document.requiredObject("shard") != expectedShard) {
            throw FullTreeDataTruthException("data observation bindings do not match")
        }

        val globals = document.requiredArray("globals").objects("data observation global")
        val types = document.requiredArray("types").objects("data observation type")
        requireOrderedById(globals, "globals")
        requireOrderedById(types, "types")
        val unitIds = shard.units.mapTo(hashSetOf()) { it.requiredString("id") }
        if ((globals + types).any { it.requiredString("unitId") !in unitIds }) {
            throw FullTreeDataTruthException("data observation owner is outside its shard")
        }

        val members = types.flatMap { it.requiredArray("members").objects("data observation member") }
        val expectedCounts = JsonObject(
            mapOf(
                "bases" to JsonPrimitive(members.count { it.requiredString("kind") == "base" }),
                "enumerators" to JsonPrimitive(members.count { it.requiredString("kind") == "enumerator" }),
                "fields" to JsonPrimitive(members.count { it.requiredString("kind") == "field" }),
                "globals" to JsonPrimitive(globals.size),
                "scannedDies" to JsonPrimitive(document.requiredObject("counts").requiredLong("scannedDies")),
                "types" to JsonPrimitive(types.size),
                "units" to JsonPrimitive(shard.units.size),
            ),
        )
        if (document.requiredObject("counts") != expectedCounts) {
            throw FullTreeDataTruthException("data observation counts do not reconcile")
        }
    }

    private fun requireOrderedById(records: List<JsonObject>, label: String) {
        val identifiers = records.map { it.requiredString("id") }
        if (identifiers != identifiers.sorted()) {
            throw FullTreeDataTruthException("data observation $label are not ordered")
        }
    }

    private fun requireSha256(value: String, label: String) {
        if (!value.matches(Regex("[0-9a-f]{64}"))) {
            throw FullTreeDataTruthException("$label digest is invalid")
        }
    }

    private val POLICY = JsonObject(
        mapOf(
            "id" to JsonPrimitive("full-tree-data-observations"),
            "version" to JsonPrimitive(6),
            "globals" to JsonPrimitive("non-declaration-static-storage-or-linkage-bearing-dwarf-variables"),
            "types" to JsonPrimitive("class-struct-union-enum-definitions-and-declarations"),
            "typeReferences" to JsonPrimitive(
                "compact-immediate-and-aggregate-dwarf-offsets-with-modifier-chain-or-closed-reason",
            ),
            "flags" to JsonPrimitive("dwarf-boolean-and-integral-forms-normalized-to-boolean"),
        ),
    )
}

internal fun JsonObject.requiredElement(name: String): JsonElement =
    this[name] ?: throw FullTreeDataTruthException("required JSON field is absent: $name")

internal fun JsonObject.requiredObject(name: String): JsonObject =
    requiredElement(name).requiredObject(name)

internal fun JsonObject.requiredArray(name: String): JsonArray =
    requiredElement(name) as? JsonArray
        ?: throw FullTreeDataTruthException("required JSON field is not an array: $name")

internal fun JsonObject.requiredString(name: String): String = requiredElement(name).requiredString(name)

internal fun JsonObject.requiredLong(name: String): Long {
    val primitive = requiredElement(name) as? JsonPrimitive
        ?: throw FullTreeDataTruthException("required JSON field is not an integer: $name")
    if (primitive.isString || primitive.booleanOrNull != null || !INTEGER_TOKEN.matches(primitive.content)) {
        throw FullTreeDataTruthException("required JSON field is not an integer: $name")
    }
    return primitive.longOrNull
        ?: throw FullTreeDataTruthException("required JSON integer is outside the supported range: $name")
}

internal fun JsonElement.requiredObject(label: String): JsonObject = this as? JsonObject
    ?: throw FullTreeDataTruthException("$label is not an object")

internal fun JsonElement.requiredString(label: String): String {
    val primitive = this as? JsonPrimitive
        ?: throw FullTreeDataTruthException("$label is not a string")
    if (!primitive.isString) throw FullTreeDataTruthException("$label is not a string")
    return primitive.content
}

internal fun JsonArray.objects(label: String): List<JsonObject> =
    mapIndexed { index, element -> element.requiredObject("$label $index") }

private val INTEGER_TOKEN = Regex("-?(?:0|[1-9][0-9]*)")
